File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    let max = 0;
    let start = 0;
    let set = {};
    for (let i = 0; i < s.length; i++) {
        let char = s[i];
        if (char in set) {
            start = Math.max(start, set[char] + 1);
        }
        max = Math.max(max, i - start + 1);
        set[char] = i;
    }
    return max;
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int right = 0;
        int ans = 0;

        HashMap<Character, Integer> map = new HashMap<>();
        while(right < s.length()) {
            if(map.containsKey(s.charAt(right))) {
                left = Math.max(map.get(s.charAt(right)) + 1, left);
            }

            map.put(s.charAt(right), right);
            ans = Math.max(ans, right - left + 1);
            right++;
        }
        return ans;
    }
}